File: 107. Binary Tree Level Order Traversal II/107. Binary Tree Level Order Traversal II.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
def level_order_bottom(root)
    return [] if root.nil?
    
    result = []
    queue = [root]
    
    while !queue.empty?
        level_size = queue.size
        current_level = []
        
        level_size.times do
            node = queue.shift
            current_level << node.val
            
            queue << node.left if node.left
            queue << node.right if node.right
        end
        
        result.unshift(current_level)
    end
    
    result
end

File: 107. Binary Tree Level Order Traversal II/107. Binary Tree Level Order Traversal II.js
function levelOrderBottom(root) {
    if (root === null) return [];

    let result = [];
    let queue = [root];

    while (queue.length > 0) {
        let levelSize = queue.length;
        let levelValues = [];

        for (let i = 0; i < levelSize; i++) {
            let node = queue.shift();  // Dequeue the front node
            levelValues.push(node.val); // Store the value of the node

            // Enqueue the children of the current node
            if (node.left) queue.push(node.left);
            if (node.right) queue.push(node.right);
        }

        result.push(levelValues); // Store the values of the current level
    }

    // Reverse the result to get the bottom-up order
    return result.reverse();
}

// Definition for a binary tree node
function TreeNode(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
}